openfl-tutorials/experimental/Numpy_Linear_Regression_Workflow.ipynb

```diff
@@ -0,0 +1,465 @@
+    "\n",
+    "    return x, y\n",
+    "\n",
+    "class SyntheticFederatedDataset:\n",
+    "    def __init__(self, num_collaborators=2, batch_size=1, num_samples=10, **kwargs):\n",
+    "        self.batch_size = batch_size\n",
+    "        X, y = generate_synthetic(rank=42, n_samples=num_samples, noise=0.1)\n",
+    "        X = np.array(X, dtype=np.float32)\n",
+    "        y = np.array(y, dtype=np.float32)\n",
+    "\n",
+    "        self.X_train_all = X[:int(0.8 * len(X))]\n",
+    "        self.y_train_all = y[:int(0.8 * len(y))]\n",
+    "        \n",
+    "        self.X_test_all = X[int(0.8 * len(X)):]\n",
+    "        self.y_test_all = y[int(0.8 * len(y)):]\n",
+    "\n",
+    "        min_samples = max(len(self.X_train_all) // num_collaborators, 1)\n",
+    "\n",
+    "        while len(self.X_test_all) < num_collaborators or any(len(chunk) < 1 for chunk in self.X_train_all):\n",
+    "            X, y = generate_synthetic(rank=42, n_samples=len(self.X_train_all) + len(self.X_valid_all) + 1, noise=0.1)\n",
+    "            X = np.array(X, dtype=np.float32)\n",
+    "            y = np.array(y, dtype=np.float32)\n",
+    "            self.X_train_all = X[:int(0.9 * len(X))]\n",
+    "            self.X_test_all = X[int(0.9 * len(X)):]\n",
+    "            self.y_train_all = y[:int(0.9 * len(y))]\n",
+    "            self.y_test_all = y[int(0.9 * len(y)):]\n",
+    "\n",
+    "        self.X_train_all = np.array_split(self.X_train_all, num_collaborators)\n",
+    "        self.X_test_all = np.array_split(self.X_test_all, num_collaborators)\n",
+    "        self.y_train_all = np.array_split(self.y_train_all, num_collaborators)\n",
+    "        self.y_test_all = np.array_split(self.y_test_all, num_collaborators)\n",
+    "\n",
+    "    def split(self, index):\n",
+    "        return {\n",
+    "            \"train_loader\":\n",
+    "                data.DataLoader(\n",
+    "                    data.TensorDataset(\n",
+    "                        pt.from_numpy(self.X_train_all[index]),\n",
+    "                        pt.from_numpy(self.y_train_all[index])\n",
+    "                    ),\n",
+    "                    batch_size=self.batch_size, shuffle=True\n",
+    "                ),\n",
+    "            \"test_loader\":\n",
+    "                data.DataLoader(\n",
+    "                    data.TensorDataset(\n",
+    "                        pt.from_numpy(self.X_test_all[index]),\n",
+    "                        pt.from_numpy(self.y_test_all[index])\n",
+    "                    ),\n",
+    "                    batch_size=self.batch_size, shuffle=True\n",
+    "                )\n",
+    "        }"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "aggregator = Aggregator()\n",
+    "aggregator.private_attributes = {}\n",
+    "\n",
+    "collaborator_names = ['Coll-1', 'Coll-2', 'Coll-3', 'Coll-4']\n",
```

Comment:
In Workflow API, there's a tradition to name collaborators after cities. Pick the ones you like (ideally geo-distributed) 😉

Reply:
Done.